src/crimson/osd/pg_shard_manager.h

```diff
@@ -54,22 +54,53 @@ class PGShardManager {
     crimson::os::FuturizedStore &store);
   seastar::future<> stop();
 
-  auto &get_osd_singleton_state() { return *osd_singleton_state; }
-  auto &get_osd_singleton_state() const { return *osd_singleton_state; }
-  auto &get_shard_services() { return *shard_services; }
-  auto &get_shard_services() const { return *shard_services; }
-  auto &get_local_state() { return shard_services->local_state; }
-  auto &get_local_state() const { return shard_services->local_state; }
+  auto &get_osd_singleton_state() {
+    ceph_assert(seastar::this_shard_id() == PRIMARY_CORE);
+    return osd_singleton_state.local();
+  }
+  auto &get_osd_singleton_state() const {
+    ceph_assert(seastar::this_shard_id() == PRIMARY_CORE);
+    return osd_singleton_state.local();
+  }
+  auto &get_shard_services() {
+    ceph_assert(seastar::this_shard_id() == PRIMARY_CORE);
+    return shard_services.local();
+  }
+  auto &get_shard_services() const {
+    ceph_assert(seastar::this_shard_id() == PRIMARY_CORE);
+    return shard_services.local();
+  }
+  auto &get_local_state() { return get_shard_services().local_state; }
+  auto &get_local_state() const { return get_shard_services().local_state; }
 
   seastar::future<> update_map(local_cached_map_t &&map) {
-    auto fmap = make_local_shared_foreign(std::move(map));
-    get_osd_singleton_state().update_map(fmap);
-    get_local_state().update_map(std::move(fmap));
-    return seastar::now();
+    get_osd_singleton_state().update_map(
+      make_local_shared_foreign(local_cached_map_t(map))
+    );
```

Comment:
so for Primary core,  osd_singleton_state and perShardState both will do update_map, right?

Reply:
Right